nltk/stem/cistem.py

```diff
@@ -0,0 +1,216 @@
+
+    In the paper, we conducted an analysis of publicly available stemmers,
+    developed two gold standards for German stemming and evaluated the stemmers
+    based on the two gold standards. We then proposed the stemmer implemented here
+    and show that it achieves slightly better f-measure than the other stemmers and
+    is thrice as fast as the Snowball stemmer for German while being about as fast
+    as most other stemmers.
+
+    :param case_insensitive: if True, the stemming is case insensitive. False by default.
+    :type case_insensitive: bool
+    """
+
+    def __init__(self, case_insensitive=False):
+        self._case_insensitive = case_insensitive
+        self._strip_ge = re.compile(r"^ge(.{4,})")
+        self._repl_xx = re.compile(r"(.)\1")
+        self._repl_xx_back = re.compile(r"(.)\*")
+        self._strip_emr = re.compile(r"e[mr]$")
+        self._strip_nd = re.compile(r"nd$")
+        self._strip_t = re.compile(r"t$")
+        self._strip_esn = re.compile(r"[esn]$")
+
+    def stem(self, word):
+        """
+        This method takes the word to be stemmed and a boolean specifiying if
+        case-insensitive stemming should be used and returns the stemmed word. If only
+        the word is passed to the method or the second parameter is 0, normal
+        case-sensitive stemming is used, if the second parameter is 1,
+        case-insensitive stemming is used.
+
+        Case sensitivity improves performance only if words in the text may be
+        incorrectly upper case. For all-lowercase and correctly cased text, best
+        performance is achieved byusing the case-sensitive version.
+
+        :param word: the word that is to be stemmed
+        :type word: unicode
+        :return word: the stemmed word
+        :rtype: unicode
+
+        >>> from nltk.stem.cistem import Cistem
+        >>> stemmer = Cistem()
+        >>> s1 = "Speicherbehältern"
+        >>> stemmer.stem(s1)
+        'speicherbehalt'
+        >>> s2 = "Grenzpostens"
+        >>> stemmer.stem(s2)
+        'grenzpost'
+        >>> s3 = "Ausgefeiltere"
+        >>> stemmer.stem(s3)
+        'ausgefeilt'
+        >>> stemmer = Cistem(True)
+        >>> stemmer.stem(s1)
+        'speicherbehal'
+        >>> stemmer.stem(s2)
+        'grenzpo'
+        >>> stemmer.stem(s3)
+        'ausgefeil'
+        """
+        if len(word) == 0:
+            return word
+
+        upper = word[0].isupper()
+        word = word.lower()
+
+        word = word.replace("ü", "u")
+        word = word.replace("ö", "o")
+        word = word.replace("ä", "a")
+        word = word.replace("ß", "ss")
+
+        word = self._strip_ge.sub(r"\1", word)
+        word = word.replace("sch", "$")
+        word = word.replace("ei", "%")
+        word = word.replace("ie", "&")
+        word = self._repl_xx.sub(r"\1*", word)
+
+        while len(word) > 3:
+            if len(word) > 5:
+                (word, success) = self._strip_emr.subn("", word)
```

Comment:
It would be more pythonic to raise an error inside the function and catch it in the outer loop here

Reply:
While I agree that the continue statements are not elegant, I think it would give the wrong idea to raise errors here because nothing has actually gone wrong. The continue statements essentially form a glorified if-statement that goes through the regexes and stops as soon as one of them has matched.